modules/hagrid.nix

```diff
@@ -0,0 +1,154 @@
+{ config, lib, pkgs, ... }:
+
+let
+  inherit (builtins) readFile toString;
+  inherit (lib) attrsets generators modules options strings types;
+  inherit (attrsets) mapAttrsToList recursiveUpdate;
+  inherit (generators) toINI;
+  inherit (modules) mkIf;
+  inherit (options) literalExpression mdDoc mkEnableOption mkOption;
+  inherit (strings) concatStrings replaceStrings;
+
+  cfg = config.services.hagrid;
+  toml = pkgs.formats.toml {};
+  defaultSettings = {
+    global = {
+      address = "0.0.0.0";
+      port = 8080;
+    };
+    release = {
+      assets_dir = "${cfg.webRoot}/assets";
+      base-URI = "https://${cfg.hostname}";
+      base-URI-Onion = "https://${cfg.hostname}";
+      email_template_dir = "${cfg.dataDir}/email-templates";
+      enable_prometheus = false;
+      from = "noreply@localhost";
+      keys_external_dir = "${cfg.webRoot}/keys";
+      keys_internal_dir = "${cfg.dataDir}/keys";
+      mail_rate_limit = 3600;
+      maintenance_file = "${cfg.dataDir}/maintenance";
+      template_dir = "${cfg.dataDir}/templates";
+      tmp_dir = "${cfg.dataDir}/tmp";
+      token_dir = "${cfg.dataDir}/tokens";
+      token_secret = "$TOKEN_SECRET";
+      token_validity = 3600;
+      x-accel-redirect = true;
+    };
+  };
+in {
+  meta.maintainers = with lib.maintainers; [ ];
+
+  options = {
+
+    services.hagrid = {
+
+      enable = mkEnableOption (mdDoc ''
+        hagrid (Verifying OpenPGP keyserver, written in Rust)'');
+
+      package = mkOption {
+        default = pkgs.hagrid;
+        defaultText = literalExpression "pkgs.hagrid";
+        type = types.package;
+        description = mdDoc "Which hagrid derivation to use.";
+      };
+
+      dataDir = mkOption {
+        type = types.path;
+        default = "/var/lib/hagrid";
+        example = "/var/db/hagrid";
+        description = mdDoc ''
+          Data directory for hagrid, where the keys, assets,
+          templates and Rocket.toml are located.
+        '';
+      };
+
+      environmentFile = mkOption {
+        type = types.path;
```

Comment:
Please use `types.str` here. If we ever pass a native Nix path in here, we'll risk leaking the secrets into the store (depending on how exactly this is used inside NixOS).